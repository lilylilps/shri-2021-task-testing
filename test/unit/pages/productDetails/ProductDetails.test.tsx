import React from "react";
import renderer from 'react-test-renderer';
import { Router } from "react-router";
import { Provider } from "react-redux";
import { createMemoryHistory } from "history";

import { it, describe, expect } from "@jest/globals";
import { render } from "@testing-library/react";
import events from "@testing-library/user-event";


import { initStore } from "../../../../src/client/store";
import { ProductDetails } from "../../../../src/client/components/ProductDetails";
import { ExampleApi, CartApi } from "../../../../src/client/api";

import { Store } from "redux";
import { Product } from "../../../../src/common/types";

describe('проверка productDetails', () => {
    let api: ExampleApi;
    let cart: CartApi;
    let store: Store;
    let product: Product;

    beforeAll(() => {
        api = new ExampleApi('');
        cart = new CartApi();
        product = {
            id: 1,
            name: "shorts",
            price: 200,
            description: 'super shorts',
            material: 'denim',
            color: 'blue',
        };
    });

    beforeEach(() => {
        cart.setState(null);
        store = initStore(api, cart);
    });

    it('snapshot', () => {
        const tree = renderer.create(
            <Provider store={store}>
                <ProductDetails product={product} />
            </Provider>)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('по клику на Add to cart появляется "Item in cart" и меняется количество товаров в корзине', () => {
        const history = createMemoryHistory({
            initialEntries: ['/catalog/1'],
            initialIndex: 0
        });

        const productPage = (
            <Router history={history}>
                <Provider store={store}>
                    <ProductDetails product={product} />
                </Provider>
            </Router>
        );

        const { getByRole, queryByText } = render(productPage);

        events.click(getByRole('button', { name: /add to cart/i }));

        let cartList = cart.getState();
        let count = cartList[1].count;
        expect(count).toEqual(1);

        const cartBage = queryByText('Item in cart');
        expect(cartBage).not.toBeNull();

        events.click(getByRole('button', { name: /add to cart/i }));
        cartList = cart.getState();
        count = cartList[1].count;
        
        expect(count).toEqual(2);
    });
});