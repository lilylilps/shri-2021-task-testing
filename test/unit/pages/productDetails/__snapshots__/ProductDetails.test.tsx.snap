// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`проверка productDetails snapshot 1`] = `
<div
  className="ProductDetails row"
>
  <div
    className="col-12 col-sm-5 col-lg-4"
  >
    <img
      className="Image"
      src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNkMAYAADkANVKH3ScAAAAASUVORK5CYII="
    />
  </div>
  <div
    className="col-12 col-sm-7 col-lg-6"
  >
    <h1
      className="ProductDetails-Name"
    >
      shorts
    </h1>
    <p
      className="ProductDetails-Description"
    >
      super shorts
    </p>
    <p
      className="ProductDetails-Price fs-3"
    >
      $
      200
    </p>
    <p>
      <button
        className="ProductDetails-AddToCart btn btn-primary btn-lg"
        onClick={[Function]}
      >
        Add to Cart
      </button>
    </p>
    <dl>
      <dt>
        Color
      </dt>
      <dd
        className="ProductDetails-Color text-capitalize"
      >
        blue
      </dd>
      <dt>
        Material
      </dt>
      <dd
        className="ProductDetails-Material text-capitalize"
      >
        denim
      </dd>
    </dl>
  </div>
</div>
`;
